core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/CqlVectorCodec.java

```diff
@@ -0,0 +1,105 @@
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.type.codec;
+
+import com.datastax.oss.driver.api.core.ProtocolVersion;
+import com.datastax.oss.driver.api.core.data.CqlVector;
+import com.datastax.oss.driver.api.core.type.CqlVectorType;
+import com.datastax.oss.driver.api.core.type.DataType;
+import com.datastax.oss.driver.api.core.type.codec.TypeCodec;
+import com.datastax.oss.driver.api.core.type.reflect.GenericType;
+import com.datastax.oss.driver.shaded.guava.common.base.Preconditions;
+import com.datastax.oss.driver.shaded.guava.common.base.Splitter;
+import com.datastax.oss.driver.shaded.guava.common.collect.Iterables;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+
+public class CqlVectorCodec implements TypeCodec<CqlVector> {
+
+  @NonNull
+  @Override
+  public GenericType<CqlVector> getJavaType() {
+    return GenericType.of(CqlVector.class);
+  }
+
+  /* Since we've overridden accepts() this shouldn't ever actually be used */
+  @NonNull
+  @Override
+  public DataType getCqlType() {
+    return new CqlVectorType(0);
+  }
+
+  @NonNull
+  @Override
+  public boolean accepts(@NonNull DataType cqlType) {
+    Preconditions.checkNotNull(cqlType);
+    return cqlType.getClass().equals(CqlVectorType.class);
+  }
+
+  @Nullable
+  @Override
+  public ByteBuffer encode(@Nullable CqlVector value, @NonNull ProtocolVersion protocolVersion) {
+    if (value == null) {
+      return null;
+    }
+    float[] values = value.getValues();
+    ByteBuffer bytes = ByteBuffer.allocate(4 * values.length);
+    for (int i = 0; i < values.length; ++i) bytes.putFloat(values[i]);
+    bytes.rewind();
+    return bytes;
+  }
+
+  @Nullable
+  @Override
+  public CqlVector decode(@Nullable ByteBuffer bytes, @NonNull ProtocolVersion protocolVersion) {
+    if (bytes == null || bytes.remaining() == 0) {
+      return null;
+    }
+    int length = bytes.limit();
+    if (length % 4 != 0)
+      throw new IllegalArgumentException("Expected CqlVector to consist of a multiple of 4 bytes");
+    int valuesCnt = length / 4;
+    float[] values = new float[valuesCnt];
+    for (int i = 0; i < valuesCnt; ++i) {
+      values[i] = bytes.getFloat();
+    }
+    /* Restore the input ByteBuffer to its original state */
+    bytes.rewind();
+    return new CqlVector(values);
+  }
```

Comment:
Note that encode/decode above represent something of a compromise.  There's no correlation between the number of floats we try to read or write and the number that are actually expected according to the type.  We do extract the expected number of dimensions for a given vector and we do store it in a CqlVectorType instance but that information isn't available to the encode/decode impls above.  This isn't a huge surprise; in the past parameterized types have been used exclusively to indicate subtypes, not meta-information about a type.

The really robust answer is probably to expand this API so that the underlying type is provided to encode() and decode().  In most cases that type will be ignored, but in this case we'd use it to see how many floats we should expect to read and write.  In the interest of time I avoided this larger-scale refactoring and settled on the compromise above.  I'm open to a follow-up ticket which would do things the right way.

Reply:
The above statement is no longer true with the changes made to support CEP-7 syntax.